src/services/shop/shop-router.ts

```diff
@@ -42,148 +43,423 @@ shopRouter.get(
-                description: "The item already exists",
-                schema: ShopItemAlreadyExistsErrorSchema,
+            [StatusCode.ClientErrorBadRequest]: {
+                description: "Not enough quantity in shop",
+                schema: ShopInsufficientQuantityErrorSchema,
+            },
+            [StatusCode.ClientErrorPaymentRequired]: {
+                description: "Not enough points to purchase",
+                schema: ShopInsufficientFundsErrorSchema,
+            },
+            [StatusCode.ServerErrorInternal]: {
+                description: "Errors that should never happen",
+                schema: ShopInternalErrorSchema,
             },
         },
     }),
     async (req, res) => {
-        const details = req.body;
-        const itemId = "item" + parseInt(crypto.randomBytes(Config.SHOP_BYTES_GEN).toString("hex"), 16);
-        const instances = Array.from({ length: details.quantity }, (_, index) => getRand(index));
+        try {
+            const { userId } = req.body;
 
-        const shopItem: ShopItem = {
-            ...details,
-            itemId: itemId,
-            instances: instances,
-        };
+            // Retrieve the user's order
+            const order = await Models.ShopOrder.findOne({ userId });
+            if (!order) {
+                return res.status(StatusCode.ServerErrorInternal).send(ShopInternalError);
+            }
 
-        // Ensure that item doesn't already exist before creating
-        const itemExists = (await Models.ShopItem.findOne({ name: details.name })) ?? false;
-        if (itemExists) {
-            return res.status(StatusCode.ClientErrorConflict).send(ShopItemAlreadyExistsError);
-        }
+            // Retrieve the user's profile
+            const profile = await Models.AttendeeProfile.findOne({ userId: order.userId });
+            if (!profile) {
+                return res.status(StatusCode.ServerErrorInternal).send(ShopInternalError);
+            }
+
+            let totalPointsRequired = 0;
+
+            // Loop through items and check availability and price
+            for (let i = 0; i < order.items.length; i++) {
+                const item = await Models.ShopItem.findOne({ itemId: order.items[i] });
+                if (!item) {
+                    return res.status(StatusCode.ClientErrorNotFound).send(ShopItemNotFoundError);
+                }
+
+                const quantity = order.quantity?.[i] ?? 0; // Default to 0 if undefined
+                totalPointsRequired += quantity * item.price;
 
-        const newItem = await Models.ShopItem.create(shopItem);
-        const withoutInstances = {
-            ...newItem.toObject(),
-            instances: undefined,
-        };
+                // Check if requested quantity is available
+                if (quantity > item.quantity) {
+                    return res.status(StatusCode.ClientErrorBadRequest).send(ShopInsufficientQuantityError);
+                }
+            }
```

Comment:
Instead of doing this loop twice - we should make these singular operations. One operation to get all the items requested. Then, loop over and check quantity. Then, check totalPointsRequired. Finally, update all the quantities as a single operation with $inc. If that works, you then deduct the total points required.

If we do this as a separate operation for each item it will be much slower for large orders.